bigflow/dataproc.py

```diff
@@ -0,0 +1,358 @@
+import base64
+import functools
+import datetime
+import time
+import inspect
+import io
+import os
+import pathlib
+import pickle
+import random
+import re
+import string
+import subprocess
+import textwrap
+import time
+import typing
+import contextlib
+import logging
+
+from google.cloud import dataproc_v1, storage
+
+import bigflow.configuration
+import bigflow.resources
+import bigflow.commons
+
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    'PySparkJob',
+]
+
+DEFAULT_REQUIREMENTS = [
+    'bigflow[dataproc,log,bigquery]==1.0.dev67',
+]
+
+
+class PySparkJob:
+
+    def __init__(
+        self,
+        id,
+        driver_callable: typing.Callable,
+        bucket_id: str,
+        gcp_project_id: str,
+        gcp_region: str,
+        driver_arguments: typing.Optional[dict] = None,
+        pip_packages: typing.Union[typing.Iterable[str], pathlib.Path] = (),
+        jar_file_uris: typing.Iterable[str] = ('gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar',),
+        worker_num_instances: int = 2,
+        worker_machine_type: str = 'n1-standard-1',
+        env: str = '',
+        setup_file: str = '',
+    ):
+        self.id = id
+
+        self.driver_filename = 'driver.py'
+        self.driver_callable = driver_callable
+        self.driver_arguments = driver_arguments or {}
+
+        self.bucket_id = bucket_id
+        self.gcp_project_id = gcp_project_id
+        self.gcp_region = gcp_region
+
+        self.env = env or bigflow.configuration.current_env() or 'none'
+
+        self.jar_file_uris = jar_file_uris
+        self.worker_machine_type = worker_machine_type
+        self.worker_num_instances = worker_num_instances
+
+        if isinstance(pip_packages, pathlib.Path):
+            self.pip_packages = bigflow.resources.read_requirements(pip_packages)
+        elif pip_packages:
+            self.pip_packages = pip_packages
+        else:
+            self.pip_packages = DEFAULT_REQUIREMENTS
+
+        if setup_file:
+            self.setup_file = setup_file
+        else:
+            self.setup_file = None
+            self._project = _capture_caller_topmodule()
+            self._any_file_inside_project = _capture_caller_path(1)
+
+    def _ensure_has_setup_file(self):
+        if self.setup_file:
+            return
+        logger.debug("Find or create setup file for main package")
```

Comment:
Can we use only 3 levels of logging - `info`, `warn`, `error`? It's the system that most of the people in Allegro use and works it works fine.